src/fragmentarium/ui/fragment/TeiExport.js

```diff
@@ -0,0 +1,117 @@
+export function teiExport(fragment) {
+  var res = ''
+  const header = getHeader(
+    fragment.number,
+    fragment.collection,
+    fragment.museum.name,
+    fragment.description,
+    fragment.publication
+  )
+  const body = getBody(fragment)
+  const end = getEnd()
+
+  res += header
+  res += body
+  res += end
+
+  console.log(fragment)
+
+  return prettifyXml(res)
+}
+
+function getHeader(title, collection, museum, sourcedesc, publication) {
+  var res = ''
+  const start =
+    '<?xml version="1.0" encoding="UTF-8"?><TEI xmlns="http://www.tei-c.org/ns/1.0"><teiHeader><fileDesc>'
+  const title_stmt = '<titleStmt><title>' + title + '</title></titleStmt>'
+  const pub_stmt =
+    '<publicationStmt><p>' +
+    museum +
+    '</p><p>' +
+    collection +
+    ' Collection</p></publicationStmt>'
+  var source_desc = '<sourceDesc><p>' + sourcedesc + '</p>'
+  if (publication) source_desc += '<p>Publication: ' + publication + '</p>'
+  const end = '</sourceDesc></fileDesc></teiHeader><text><body>'
+
+  return res + start + title_stmt + pub_stmt + source_desc + end
+}
+
+function getEnd() {
+  return '</body></text></TEI>'
+}
+
+function getBody(fragment) {
```

Comment:
The nested for-if-if construct is rather hard to read and definitely too complex. If you extract methods for the emptines check and creating the p-element, you can simplify this function to one line e.g. `fragment.text.allLines.filter(isNotEmpty).map(getParagraph).join('')`.